Renesas RX65N Support

rbateman-witekio opened the pull request:
- Add a demo application (basic-sample) for the Renesas RX65N (green Wi-Fi) board. This supports sending telemetry, receiving a command to control the LED on the cloud extension board. 
- Update setup-project.sh to support the new target (rx65ncloudkit)
- Update setup-project.sh to support Windows symbolic links (shortcuts)

rbateman-witekio committed:
Add rx65ncloudkit option to setup-project.sh (rough), add rx65x IoTHub sample projects

rbateman-witekio committed:
Add basic_sample demo application

rbateman-witekio committed:
Fix WiFi pin configuration in basic sample

rbateman-witekio committed:
Increase heap size, use IOTC gettimeofday and ctime implementations

rbateman-witekio committed:
Disable optimisations for sbrk - workaround for failing malloc

rbateman-witekio committed:
Read RX65N user switch, send as "button" value

rbateman-witekio committed:
Remove HardwareDebug artefacts directory

rbateman-witekio committed:
Update .gitignore files for rx65n samples

rbateman-witekio committed:
Add RX65N .project/.cproject files for azure rtos components

rbateman-witekio committed:
Remove MS RX65N sample_azure_iot_embedded_sdk project

rbateman-witekio committed:
Add missing sample_device_identity.c file

rbateman-witekio committed:
Rename RX65N project basic_sample -> basic-sample

rbateman-witekio committed:
Move app_config from src to include for RX65N

rbateman-witekio committed:
Add Windows symlink support to setup-project.sh

rbateman-witekio committed:
Add HardwareDebug directory to .gitignore

rbateman-witekio committed:
Remove unused demo_printf() function

rbateman-witekio committed:
Change file separator from Windows to Linux in basic-sample.scfg

rbateman-witekio committed:
Remove/fix incorrect incorrect include paths

rbateman-witekio committed:
Add LED setup code for RX65N, handle LED command

rbateman-witekio committed:
Use safe functions (stncmp/strtok_r) in on_command()

rbateman-witekio committed:
Populate 'Release' configuration with missing includes, tidy include list

nik-markovic commented:
I see that there are third party files which have been checked into the repo from azure SDK sample. Did we actually make modifications to them compared to some third party baseline? If not, would it not be too difficult to gitignore those files and import them with some automated script during the setup-project.sh?

rbateman-witekio commented:
> I see that there are third party files which have been checked into the repo from azure SDK sample. Did we actually make modifications to them compared to some third party baseline? If not, would it not be too difficult to gitignore those files and import them with some automated script during the setup-project.sh?

Hi Nik, I assume you are talking about the stuff in smc_gen (all the boilerplate BSP code)? I guess we could just not commit them at all and let the tool generate it at build time. There is one file in particular there which we have changed. However, it might be that if we just commit that one file e2 studio will just generate the missing artefacts. I'll give it a go here and update accordingly.

nik-markovic commented on scripts/setup-project.sh:
```diff
@@ -126,23 +134,33 @@ case "$name" in
+    ;;
+  rx65ncloudkit)
+    echo Downloading Azure_RTOS_6...
+	  wget -q -O azrtos.zip https://github.com/azure-rtos/samples/releases/download/v6.1_rel/Azure_RTOS_6.1_RX65N_Cloud_Kit_E2Studio_GNURX_Samples_2022_05_25.zip
```
AzureRTOS had deleted a release (zip) once in the past and messed up our build. Since then we have been uploading a copy that's hosted by our own storage. Use the newly uploaded zip at  https://saleshosted.blob.core.windows.net/$web/sdk/AzureRTOS/Azure_RTOS_6.1_RX65N_Cloud_Kit_E2Studio_GNURX_Samples_2022_05_25.zip

nik-markovic commented:
> 
> > I see that there are third party files which have been checked into the repo from azure SDK sample. Did we actually make modifications to them compared to some third party baseline? If not, would it not be too difficult to gitignore those files and import them with some automated script during the setup-project.sh?
> 
> Hi Nik, I assume you are talking about the stuff in smc_gen (all the boilerplate BSP code)? I guess we could just not commit them at all and let the tool generate it at build time. There is one file in particular there which we have changed. However, it might be that if we just commit that one file e2 studio will just generate the missing artefacts. I'll give it a go here and update accordingly.

Yes, I believe only the files in smc_gen directory. I've seen some Azure SDK sample files checked in at some point, but I think those have been removed from the tree at some point.

nik-markovic commented on samples/rx65ncloudkit/basic_sample/src/iotconnect_app.c:
```diff
@@ -13,6 +13,9 @@
+
 // from nx_azure_iot_adu_agent_<boardname>_driver.c
 void nx_azure_iot_adu_agent_driver(void)
 {}
```
It looks like you may have taken the unlucky version of iotconnect_app.c as a baseline. The ADU driver with no implementation was wrong in the original code. See PR #18. If you don't support OTA, you can remove this declaration completely.

rbateman-witekio commented:
> Yes, I believe only the files in smc_gen directory. I've seen some Azure SDK sample files checked in at some point, but I think those have been removed from the tree at some point.

You are right, I copied the Azure SDK example directly into the project as a starting point. Then removed it once I had the basic-sample project working. These commits should get squashed upon merging right?

nik-markovic commented on samples/rx65ncloudkit/basic_sample/src/iotconnect_app.c:
```diff
@@ -0,0 +1,358 @@
+        } else if (app_needs_ota_update(version)) {
+            printf("OTA update is required for version %s.\r\n", version);
+
+            if (start_ota(url)) {
```
It looks like you took an OTA capable version iotconnect_app.c. The non-OTA-capable boilerplate in  iotconnect_app doesn't try to attempt to start the OTA and returns *message = "Not implemented"*

nik-markovic commented on scripts/setup-project.sh:
```diff
@@ -109,20 +131,17 @@ case "$name" in
     pushd iotc-azrtos-sdk/ >/dev/null
       for f in ../../../iotc-azrtos-sdk/*; do
-        ln -sf $f .
+        make_sym_link $f $(get_file_name $f)
```
Have we actually tested opening and compiling a project under Windows for the other three projects which are in the case statement? If not, you could just do Renesas as a separate case (or an *if* check under that case) and Shu and I can start introducing make_sym_link into other projects as we validate them. Great idea BTW!

rbateman-witekio replied:
Hi Nik, I took the time today to download all of the tools and tried to compile each project under Windows with this script. The STM32L4 and MIMXRT1060 projects compile fine. 

There seems to be a build error with the OTA components in the SAME54XPro project.  It looks like the nx_azure_iot_adu_agent_driver.c file is not included in the project. So, the extern function definition in iotconnect_app.c fails to resolve. If I add the file, there are errors about missing headers for the flash driver. If I remove the start_ota function and the adu driver registration, the application builds fine. I don't think this is related to the switch to make_sym_link, I'll give it a go on "main" and see how it fairs there.

**EDIT**
Of course, it doesn't work under Windows on "main". The symlinks don't work there on Windows! Is it possible for you to confirm what I'm seeing on Linux?

nik-markovic replied:
> Hi Nik, I took the time today to download all of the tools and tried to compile each project under Windows with this script. The STM32L4 and MIMXRT1060 projects compile fine.
> 
> There seems to be a build error with the OTA components in the SAME54XPro project. It looks like the nx_azure_iot_adu_agent_driver.c file is not included in the project. So, the extern function definition in iotconnect_app.c fails to resolve. If I add the file, there are errors about missing headers for the flash driver. If I remove the start_ota function and the adu driver registration, the application builds fine. I don't think this is related to the switch to make_sym_link, I'll give it a go on "main" and see how it fairs there.
> 
> **EDIT** Of course, it doesn't work under Windows on "main". The symlinks don't work there on Windows! Is it possible for you to confirm what I'm seeing on Linux?

I got the same error on Linux. I will check with Shu. We can resolve this.

nik-markovic commented:
> removed

I squash the merge to main branch and delete the branch right away. I'm not worried about them increasing the repo size, but it's best to be clear which files are the ones that we changed, compared to some third party baseline.

The easiest way to do this is to make sure to check in the modified file, gitignore everything but the checked in file and ensure that it's done at each directory level, like at [the atec608 branch](https://github.com/avnet-iotconnect/iotc-azurertos-sdk/tree/same54x-atecc608-support/samples/same54xprov2/src).  If you try to use "gtignore all directories but specific_dir/path/file.c", it causes issues where file.c cannot be checked in properly because specific_dir/path/ directory is ignored, so that's why gitignore at each nested level.

When we override files in the pulled Azure RTOS zip, we do a *cp -n*, which doesn't overwrite. Now, id this is part of some IDE boilerplate, I hope the IDE won't try to overwrite with own generated files.

nik-markovic commented:
> > Yes, I believe only the files in smc_gen directory. I've seen some Azure SDK sample files checked in at some point, but I think those have been removed from the tree at some point.
> 
> You are right, I copied the Azure SDK example directly into the project as a starting point. Then removed it once I had the basic-sample project working. These commits should get squashed upon merging right?


I squash the merge to main branch and delete the branch right away. I'm not worried about them increasing the repo size, but it's best to be clear which files are the ones that we changed, compared to some third party baseline.

The easiest way to do this is to make sure to check in the modified file, gitignore everything but the checked in file and ensure that it's done at each directory level, like at [the atec608 branch](https://github.com/avnet-iotconnect/iotc-azurertos-sdk/tree/same54x-atecc608-support/samples/same54xprov2/src).  If you try to use "gtignore all directories but specific_dir/path/file.c", it causes issues where file.c cannot be checked in properly because specific_dir/path/ directory is ignored, so that's why gitignore at each nested level.

When we override files in the pulled Azure RTOS zip, we do a *cp -n*, which doesn't overwrite. Now, id this is part of some IDE boilerplate, I hope the IDE won't try to overwrite with own generated files.

nik-markovic commented:
I see that there are no updates to README.md for the new board. Should we do this at a later time? Also main.yml in .github directory for automated build.

I could use the instructions to validate that when the project is loaded into the IDE from version control and build, that no generated files appear as if they need to be added to git (untracked, and not gitignored). If you have done that check, I will trust it.

rbateman-witekio committed:
Remove unmodified auto generated files for RX65N

rbateman-witekio committed:
Update README.md with RX65N info

rbateman-witekio committed:
Update main.yml with RX65N artefacts

rbateman-witekio commented:
> I see that there are no updates to README.md for the new board. Should we do this at a later time? Also main.yml in .github directory for automated build.
> 
> I could use the instructions to validate that when the project is loaded into the IDE from version control and build, that no generated files appear as if they need to be added to git (untracked, and not gitignored). If you have done that check, I will trust it.

Hi Nik, I've removed the unmodified, pre-generated files. This removes a fair amount of code from the PR. I've also updated main.yml and the README file with some build instructions. Thanks for taking a look!

nik-markovic reviewed (changes requested):
Requesting some additional changes.

nik-markovic committed:
Merge branch 'main' into rx65n-support

Signed-off-by: Nik Markovic <nikola.m@gmail.com>

rbateman-witekio committed:
Remove OTA components from iotconnect_app.c in the RX65N project

rbateman-witekio committed:
Use self hosted zip file for RX65N Azure RTOS project
